src/xwrap.rs

```diff
@@ -2,340 +2,248 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-use std::ffi;
-use std::mem;
-use std::os::raw;
-use std::ptr;
-use std::slice;
-
 use image::Rgba;
 use image::RgbaImage;
-use x11::xlib;
-use x11::xrandr;
+use x11rb::connection::Connection;
+use x11rb::protocol::randr::ConnectionExt as _;
+use x11rb::protocol::xproto::{self, ConnectionExt as _};
+use x11rb::rust_connection::RustConnection;
 
 use crate::util;
 
 pub struct Display {
-    handle: *mut xlib::Display,
+    conn: RustConnection,
+    screen: usize,
 }
 
 pub struct Image {
-    handle: *mut xlib::XImage,
-}
-
-pub struct ScreenRectIter<'a> {
-    dpy: &'a Display,
-    res: *mut xrandr::XRRScreenResources,
-    crtcs: &'a [xrandr::RRCrtc],
-    i: usize,
+    w: u32,
+    h: u32,
+    format: xproto::Format,
+    visual: xproto::Visualtype,
+    byte_order: xproto::ImageOrder,
+    data: Vec<u8>,
 }
 
 impl Display {
-    pub fn open(name: Option<ffi::CString>) -> Option<Display> {
-        unsafe {
-            let name = match name {
-                None => ptr::null(),
-                Some(cstr) => cstr.as_ptr(),
-            };
-            let d = xlib::XOpenDisplay(name);
-
-            if d.is_null() {
-                None
-            } else {
-                Some(Display { handle: d })
-            }
-        }
+    pub fn open(name: Option<&str>) -> Option<Display> {
+        let (conn, screen) = x11rb::connect(name).ok()?;
+        Some(Display { conn, screen })
     }
 
-    pub fn get_default_root(&self) -> xlib::Window {
-        unsafe { xlib::XDefaultRootWindow(self.handle) }
+    fn screen(&self) -> &xproto::Screen {
+        &self.conn.setup().roots[self.screen]
```

Comment:
Seems odd that `setup` can be seemingly called multiple times.

Reply:
It's a getter for a private field. Same as calling `.len()` on a collection.